// sequelize-typescript/dist/browser/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BeforeDelete = exports.BeforeDefine = exports.BeforeCreate = exports.BeforeCount = exports.BeforeConnect = exports.BeforeBulkUpdate = exports.BeforeBulkSync = exports.BeforeBulkRestore = exports.BeforeBulkDestroy = exports.BeforeBulkDelete = exports.BeforeBulkCreate = exports.AfterValidate = exports.AfterUpsert = exports.AfterUpdate = exports.AfterSync = exports.AfterSave = exports.AfterRestore = exports.AfterInit = exports.AfterFind = exports.AfterDestroy = exports.AfterDelete = exports.AfterDefine = exports.AfterCreate = exports.AfterConnect = exports.AfterBulkUpdate = exports.AfterBulkSync = exports.AfterBulkRestore = exports.AfterBulkDestroy = exports.AfterBulkDelete = exports.AfterBulkCreate = exports.HasOne = exports.HasMany = exports.BelongsToMany = exports.BelongsTo = exports.UpdatedAt = exports.Unique = exports.Table = exports.Scopes = exports.PrimaryKey = exports.ForeignKey = exports.DeletedAt = exports.DefaultScope = exports.Default = exports.CreatedAt = exports.Comment = exports.Column = exports.AutoIncrement = exports.AllowNull = exports.DataType = exports.Model = void 0;
exports.Validator = exports.Validate = exports.NotNull = exports.NotIn = exports.NotEmpty = exports.NotContains = exports.Not = exports.Min = exports.Max = exports.Length = exports.IsUUID = exports.IsUrl = exports.IsUppercase = exports.IsNumeric = exports.IsNull = exports.IsLowercase = exports.IsIPv6 = exports.IsIPv4 = exports.IsIP = exports.IsInt = exports.IsIn = exports.IsFloat = exports.IsEmail = exports.IsDecimal = exports.IsDate = exports.IsCreditCard = exports.IsBefore = exports.IsArray = exports.IsAlphanumeric = exports.IsAlpha = exports.IsAfter = exports.Is = exports.Equals = exports.Contains = exports.ValidationFailed = exports.BeforeValidate = exports.BeforeUpsert = exports.BeforeUpdate = exports.BeforeSync = exports.BeforeSave = exports.BeforeRestore = exports.BeforeInit = exports.BeforeFindAfterOptions = exports.BeforeFindAfterExpandIncludeAll = exports.BeforeFind = exports.BeforeDestroy = void 0;
function noop() {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    return (target, propertyName) => {
        // noop
    };
}
class Model {
}
exports.Model = Model;
exports.DataType = {};
exports.AllowNull = noop;
exports.AutoIncrement = noop;
exports.Column = noop;
exports.Comment = noop;
exports.CreatedAt = noop;
exports.Default = noop;
exports.DefaultScope = noop;
exports.DeletedAt = noop;
exports.ForeignKey = noop;
exports.PrimaryKey = noop;
exports.Scopes = noop;
exports.Table = noop;
exports.Unique = noop;
exports.UpdatedAt = noop;
exports.BelongsTo = noop;
exports.BelongsToMany = noop;
exports.HasMany = noop;
exports.HasOne = noop;
exports.AfterBulkCreate = noop;
exports.AfterBulkDelete = noop;
exports.AfterBulkDestroy = noop;
exports.AfterBulkRestore = noop;
exports.AfterBulkSync = noop;
exports.AfterBulkUpdate = noop;
exports.AfterConnect = noop;
exports.AfterCreate = noop;
exports.AfterDefine = noop;
exports.AfterDelete = noop;
exports.AfterDestroy = noop;
exports.AfterFind = noop;
exports.AfterInit = noop;
exports.AfterRestore = noop;
exports.AfterSave = noop;
exports.AfterSync = noop;
exports.AfterUpdate = noop;
exports.AfterUpsert = noop;
exports.AfterValidate = noop;
exports.BeforeBulkCreate = noop;
exports.BeforeBulkDelete = noop;
exports.BeforeBulkDestroy = noop;
exports.BeforeBulkRestore = noop;
exports.BeforeBulkSync = noop;
exports.BeforeBulkUpdate = noop;
exports.BeforeConnect = noop;
exports.BeforeCount = noop;
exports.BeforeCreate = noop;
exports.BeforeDefine = noop;
exports.BeforeDelete = noop;
exports.BeforeDestroy = noop;
exports.BeforeFind = noop;
exports.BeforeFindAfterExpandIncludeAll = noop;
exports.BeforeFindAfterOptions = noop;
exports.BeforeInit = noop;
exports.BeforeRestore = noop;
exports.BeforeSave = noop;
exports.BeforeSync = noop;
exports.BeforeUpdate = noop;
exports.BeforeUpsert = noop;
exports.BeforeValidate = noop;
exports.ValidationFailed = noop;
exports.Contains = noop;
exports.Equals = noop;
exports.Is = noop;
exports.IsAfter = noop;
exports.IsAlpha = noop;
exports.IsAlphanumeric = noop;
exports.IsArray = noop;
exports.IsBefore = noop;
exports.IsCreditCard = noop;
exports.IsDate = noop;
exports.IsDecimal = noop;
exports.IsEmail = noop;
exports.IsFloat = noop;
exports.IsIn = noop;
exports.IsInt = noop;
exports.IsIP = noop;
exports.IsIPv4 = noop;
exports.IsIPv6 = noop;
exports.IsLowercase = noop;
exports.IsNull = noop;
exports.IsNumeric = noop;
exports.IsUppercase = noop;
exports.IsUrl = noop;
exports.IsUUID = noop;
exports.Length = noop;
exports.Max = noop;
exports.Min = noop;
exports.Not = noop;
exports.NotContains = noop;
exports.NotEmpty = noop;
exports.NotIn = noop;
exports.NotNull = noop;
exports.Validate = noop;
exports.Validator = noop;
